### PR TITLE
EIP55

### LINKED ISSUE #6941: [EPIC] Implement EIP-55 - Address Validity Checksum
# Problem

Users should be protected against accidental typos in Ethereum addresses

https://eips.ethereum.org/EIPS/eip-55

# Implementation

All screens that expose addresses should render it in mixed case according to eip-55
All inputs that accept addresses should display a warning if the entered address is not in mixed case or if it is not valid according to eip-55

*Update*:
Warning screen UI
Warning copy should read `Error in address`.

image

# Future Steps

Steps which should be taken after this issue has been resolved.

### PR DESCRIPTION
**Pull request by mestorlx**
Fixes #6941

### Summary:

[comment]: # (Summarise the problem and how the pull request solves it)
Implements EIP55

### Review notes:
Added the utility with tests 
Pending to integrate with the ui.

### Steps to test:
```terminal
$lein test-cljs
```

### CODE REVIEW

**Comment by mestorlx** [comment]
@StatusSceptre just submitted WIP PR.

File: src/status_im/js_dependencies.cljs [hunk 1]
```diff
@@ -9,3 +9,4 @@
 (def text-encoding       (js/require "text-encoding"))
 (def js-sha3             (js/require "js-sha3"))
 (def web3-utils          (js/require "web3-utils"))
+(def keccak              (js/require "keccak"))
```
Comment (jeluard) [inline src/status_im/js_dependencies.cljs]: Is there a way to avoid this dependency and rely on existing deps like web3 or web3-utils?

Comment (mestorlx) [reply src/status_im/js_dependencies.cljs]: @jeluard Yes, sorry!
Removed.

**Comment by mestorlx** [comment]
@jeluard if you are ok with this implementation then will add the pop-up on a test screen according to #6941

File: src/status_im/utils/ethereum/eip55.cljs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+   This EIP standardize how ethereum addresses should be printed as strings to validate checksum.
+
+   e.g. 0x5aAeb6053F3E94C9b9A09f33669435E7Ef1BeAed"
+  (:require [clojure.string :as string]
+            [status-im.utils.ethereum.core :as ethereum]))
+
+(defn verifyAddressChecksum [address]
+  "verify address checksum according to EIP 55"
+  (def adHash (ethereum/naked-address (ethereum/sha3
```
Comment (jeluard) [inline src/status_im/utils/ethereum/eip55.cljs]: It is more idiomatic to use a `let` to create local vars.

File: src/status_im/utils/ethereum/eip55.cljs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+(ns status-im.utils.ethereum.eip55
+  "Utility function related to [EIP55](https://eips.ethereum.org/EIPS/eip-55)
+
+   This EIP standardize how ethereum addresses should be printed as strings to validate checksum.
+
+   e.g. 0x5aAeb6053F3E94C9b9A09f33669435E7Ef1BeAed"
+  (:require [clojure.string :as string]
+            [status-im.utils.ethereum.core :as ethereum]))
+
+(defn verifyAddressChecksum [address]
```
Comment (jeluard) [inline src/status_im/utils/ethereum/eip55.cljs]: How about `valid-address-checksum?` ?

**Comment by jeluard** [comment]
@mestorlx LGTM? Maybe @goranjovic can comment too

File: src/status_im/utils/ethereum/eip55.cljs (hunk 2 above)
Comment (mestorlx) [reply src/status_im/utils/ethereum/eip55.cljs]: updated

File: src/status_im/utils/ethereum/eip55.cljs (hunk 3 above)
Comment (mestorlx) [reply src/status_im/utils/ethereum/eip55.cljs]: updated.

**Comment by mestorlx** [comment]
@jeluard Added the pop-up in the wallet view. Go to send, tap choose recipient and enter recipient address. If you need some address for test I recommend using the ones in the added test file.
From #6941 the UI should be different. Is that widget available or it needs to be implemented?
Also, I entered manually the string into the translation file not sure if that’s the correct procedure.

**Comment by jeluard** [comment]
> Also, I entered manually the string into the translation file not sure if that’s the correct procedure.
That's fine!

@goranjovic Can you comment on the other question from @mestorlx ?

**Comment by goranjovic** [comment]
@mestorlx 

> From #6941 the UI should be different. Is that widget available or it needs to be implemented?

It's being developed in parallel. Don't worry about that, I'll reconcile the changes when it gets merged. 

Just two more things from you needed now:
- please squash the commits
- post a screenshot please

cheers :beer:

**Comment by mestorlx** [comment]
@goranjovic Just squashed and signed the commit.
here is the screenshot of the modified screen.
let me know what else you may need from me.
Thanks


iPhone 8 simulator screen

**Comment by mestorlx** [comment]
status: ready
### Pull Request Checklist
* [x]  Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status.html))?

**Review by goranjovic — APPROVED** [review]

**Review by jeluard — APPROVED** [review]

**Comment by antdanchenko** [comment]
@goranjovic the app was adding `0x` to addresses typed without `0x`. 

This feature is gone in this PR, is that expected change?

**Comment by mestorlx** [comment]
> @goranjovic the app was adding `0x` to addresses typed without `0x`.
> 
> This feature is gone in this PR, is that expected change?

This error is due to the address not being with mixed case. Should be “0xf184747445c3B85CEb147DfB136067CB93d95F1D” will update the test files

**Comment by antdanchenko** [comment]
thanks @mestorlx

**Comment by antdanchenko** [comment]
@mestorlx thanks for taking care of tests

**Comment by StatusSceptre** [comment]
@mestorlx would you mind shooting an email to me at Chris.hutchinson@status.im so we could chat a little more about your contribution? 

Thanks for all the work gone in so far :)

**Comment by mestorlx** [comment]
> [Chris.hutchinson@status.im](mailto:Chris.hutchinson@status.im)

Sent!

**Comment by asemiankevich** [comment]
@mestorlx @jeluard where i can take the examples of valid / invalid addresses then? I checked with 

```# All caps
0x52908400098527886E0F7030069857D2E4169EE7
0x8617E340B3D01FA5F11F306F4090FD50E238070D
# All Lower
0xde709f2102306220921060314715629080e2fb77
0x27b1fdb04752bbc536007a920d24acb045561c26
# Normal
0x5aAeb6053F3E94C9b9A09f33669435E7Ef1BeAed
0xfB6916095ca1df60bB79Ce92cE3Ea74c37c5d359
0xdbF03B407c01E7cD3CBea99509d93f8DDDC8C6FB
0xD1220A0cf47c7B9Be7A2E6BA89F429762e7b9aDb
```  

from https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md but did not get any errors, maybe i understand it wrong. Please assist.

**Comment by jeluard** [comment]
@asemiankevich You can try with `0xDE709F2102306220921060314715629080E2fB77`

**Comment by mestorlx** [comment]
> @mestorlx @jeluard where i can take the examples of valid / invalid addresses then? I checked with
> 
> ```
> 0x52908400098527886E0F7030069857D2E4169EE7
> 0x8617E340B3D01FA5F11F306F4090FD50E238070D
> # All Lower
> 0xde709f2102306220921060314715629080e2fb77
> 0x27b1fdb04752bbc536007a920d24acb045561c26
> # Normal
> 0x5aAeb6053F3E94C9b9A09f33669435E7Ef1BeAed
> 0xfB6916095ca1df60bB79Ce92cE3Ea74c37c5d359
> 0xdbF03B407c01E7cD3CBea99509d93f8DDDC8C6FB
> 0xD1220A0cf47c7B9Be7A2E6BA89F429762e7b9aDb
> ```
> from https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md but did not get any errors, maybe i understand it wrong. Please assist.

@asemiankevich These address are all valid. Try replacing any letter capitalization of these address and that should fail.

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/ethereum/EIPs">ethereum/EIPs</a></strong></div><div>The Ethereum Improvement Proposal repository. Contribute to ethereum/EIPs development by creating an account on GitHub.</div></blockquote>

**Comment by asemiankevich** [comment]
@mestorlx @jeluard thanks!

**Comment by jeluard** [comment]
@mestorlx Thanks for your contribution! Keep them coming! :+1:

**Pull request merged**